### DISCUSSION TITLE
A related tool for Zod schemas

### DISCUSSION THREAD

**Discussion by Southclaws** [gzuidhof/tygo#1] [Show and tell]
Hey! A friend shared this in my community and it looks useful as I usually use Go and TypeScript together!

Figured it was a good place to plug my own related tool which is specifically for Zod schemas in TS: https://github.com/Southclaws/supervillain/

**Comment by gzuidhof** [comment]
That's really cool! I'm a big fan of Zod (somewhat long-time sponsor too :)).

I've thought about having a more data-intermediate approach to generation which would be possible to be used to output in different formats and perform resolving of imports too. Right now it outputs code as it goes through the AST without an intermediate representation. So for a statement like

```go
/* some godoc */
type Foo = map[string]bar.Dep // line comment
```

It would generate some sort of intermediate representation
```golang
TypeDecl {
  Doc: "some godoc",
  Name: "Foo"
  Type: MapDecl {
       ....
  }
  Comment: "line comment"
}
```

From that data it would then be somewhat easy to serialize into other structures like Zod, as well as to be a building block for more complex usecases like generating an API client entirely (you would put some sort of directives into comments, like `@tygo_path "/v1/my/endpoint"`).

Thinking about it a bit more: I'm not sure what you are using Zod for in this case? I've used it on the backend to validate user inputs, but with a Golang backend I'm not sure where it fits (unless the user is directly inputting JSON into a textfield).

**Reply by Southclaws** [reply]
I use it as a sanity check to make sure the payloads my API is sending are exactly what I expect on the frontend. It's one step further than a blind `await response.json() as SomeType` as it performs validation too. So if a developer lazily implements a frontend feature and uses the wrong type, they'll get a validation error and be forced to use the correct type. If you accidentally use the wrong type in an `as` type assertion then you don't find out until you attempt to use the payload, much later in the program (usually resulting in a confusion "cannot read x of y/undefined" and lots of debugging, depending on how large your app is)

I apply schemas at every entrypoint of external data: forms/user input and API calls - even if that API is our own API, I still treat it as "external" and thus put it through the same level of validation.

**Comment by gzuidhof** [comment]
I was Googling a bit, you could chain this in the pipeline https://www.npmjs.com/package/ts-to-zod as the comments are persisted
